### PR TITLE
Dictionary compression

### PR DESCRIPTION
**Pull request by bobbieltd**
If you have time, add this to benchmark. I'm testing this with Wrkzd on my server.

### CODE REVIEW

**Comment by wrkzdev** [comment]
Thanks you and will do. Any better result, I will update in the existing issue.

**Comment by wrkzdev** [comment]
In term of data dir size,  I would say no different.

```
Complete sync: 1h 35m 56s with checkpoints file.
Height: 604255/604255 (100.00%) on mainnet, synced, net hash 3.02 MH/s, v6, up to date, 8(out)+10(in) connections, uptime 0d 1h 35m 56s

du -h data
9.3G    data/DB
14G     data
```
Comparing to no compression up-to-date same height:
```
du -h data
14G     data/DB
18G     data
```

**Comment by wrkzdev** [comment]
Testing same machine as this:
https://github.com/wrkzcoin/wrkzcoin/issues/4

Wallet sync result with zedwallet-beta (default thread 8) run 1) then 2) after.

| Wallet              	| Num. Blocks 	| Duration 	| Wallet tx    	| Blocks/s 	|
|---------------------	|-------------	|----------	|--------------	|----------	|
| 1) Non-empty Wallet 	|   604,331   	| 1h11mn   	| 30,492 tx    	|   ~142   	|
| 2) Empty Wallet     	|   604,359   	| 22mn     	| Empty Wallet 	|    457   	|

**Comment by bobbieltd** [comment]
Thanks for benchmark. It does not improve on my server too. Maybe, it can not make dictionary with arbitrary data.

File: src/CryptoNoteCore/RocksDBWrapper.cpp [hunk 1]
```diff
@@ -52,6 +52,12 @@ void RocksDBWrapper::init(const DataBaseConfig& config) {
   rocksdb::DB* dbPtr;
 
   rocksdb::Options dbOptions = getDBOptions(config);
+  if (config.getCompressionEnabled()) {
+      dbOptions.compression_opts.strategy = 9; // level 9
```
Comment (wrkzdev) [inline src/CryptoNoteCore/RocksDBWrapper.cpp]: What we missed is .level :)

Comment (bobbieltd) [reply src/CryptoNoteCore/RocksDBWrapper.cpp]: Thanks. I think I missed also the dict part at column setting below. Settings here won't affect.

Comment (wrkzdev) [reply src/CryptoNoteCore/RocksDBWrapper.cpp]:
```
Process ./data/DB/000027.sst
Sst file format: block-based
Table Properties:
------------------------------
  # data blocks: 16803
  # entries: 296259
  # deletions: 0
  # merge operands: 0
  # range deletions: 0
  raw key size: 12739137
  raw average key size: 43.000000
  raw value size: 121066608
  raw average value size: 408.651241
  data block size: 107378955
  index block size (user-key? 0, delta-value? 0): 915999
  filter block size: 0
  (estimated) table size: 108294954
  filter policy name: N/A
  prefix extractor name: nullptr
  column family ID: 0
  column family name: default
  comparator name: leveldb.BytewiseComparator
  merge operator name: nullptr
  property collectors names: []
  SST file compression algo: ZSTD
  SST file compression options: window_bits=-14; level=16; strategy=0; max_dict_bytes=16384; zstd_max_train_bytes=262144; enabled=1;
  creation time: 1567296028
  time stamp of earliest key: 0
  file creation time: 1567296054
  Raw user collected properties
------------------------------
  # rocksdb.block.based.table.index.type: 0x00000000
  # rocksdb.block.based.table.prefix.filtering: 0x30
  # rocksdb.block.based.table.whole.key.filtering: 0x31
  # rocksdb.deleted.keys: 0x00
  # rocksdb.merge.operands: 0x00
```
One output from sst_dump. I adjust the `max_dict_bytes` and `zstd_max_train_bytes`. Still doing some process of testings :)

Comment (bobbieltd) [reply src/CryptoNoteCore/RocksDBWrapper.cpp]: I think I misconfigure dict with this commit.
It should be :
    // zstd compression settings
    if (config.getCompressionEnabled()) {
        fOptions.compression_opts.level = 9; // level 9
        fOptions.compression_opts.strategy = 9; // level 9
        fOptions.compression_opts.max_dict_bytes = 1 << 16;        // 64KB
        fOptions.compression_opts.zstd_max_train_bytes = 1 << 20;  // 1024KB
        fOptions.compression_opts.enabled = true;
    }

I'm testing it with wrkz

Comment (bobbieltd) [reply src/CryptoNoteCore/RocksDBWrapper.cpp]: Ouff, bad news. With new dictionary setting, compression ratio is much worse.

**Pull request closed without merging**